object-handlers.go

```diff
@@ -600,75 +629,20 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 			return
 		}
 		// Create anonymous object.
-		md5Sum, err = api.ObjectAPI.PutObject(bucket, object, size, r.Body, metadata)
+		md5Sum, err = api.ObjectAPI.PutObject(bucket, object, size, r.Body, metadata, nil)
 	case authTypePresigned, authTypeSigned:
-		validateRegion := true // Validate region.
-
-		if skipSHA256Calculation(r) {
-			// Either sha256-header is "UNSIGNED-PAYLOAD" or this is a presigned PUT
-			// request without sha256-header.
-			var s3Error APIErrorCode
-			if isRequestSignatureV4(r) {
-				s3Error = doesSignatureMatch(unsignedPayload, r, validateRegion)
-			} else if isRequestPresignedSignatureV4(r) {
-				s3Error = doesPresignedSignatureMatch(unsignedPayload, r, validateRegion)
-			}
-			if s3Error != ErrNone {
-				if s3Error == ErrSignatureDoesNotMatch {
-					err = errSignatureMismatch
-				} else {
-					err = fmt.Errorf("%v", getAPIError(s3Error))
-				}
-			} else {
-				md5Sum, err = api.ObjectAPI.PutObject(bucket, object, size, r.Body, metadata)
-			}
+		var signVerifyFn signVerifyFunc
+		var reader io.Reader
+		if !skipSHA256Calculation(r) {
+			shaWriter := sha256.New()
+			reader = io.TeeReader(r.Body, shaWriter)
+			signVerifyFn = signVerifyFactory(shaWriter, r)
 		} else {
-			// Sha256 of payload has to be calculated and matched with what was sent in the header.
-			// Initialize a pipe for data pipe line.
-			reader, writer := io.Pipe()
-			var wg = &sync.WaitGroup{}
-			// Start writing in a routine.
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				shaWriter := sha256.New()
-				multiWriter := io.MultiWriter(shaWriter, writer)
-				if _, wErr := io.CopyN(multiWriter, r.Body, size); wErr != nil {
-					// Pipe closed.
-					if wErr == io.ErrClosedPipe {
-						return
-					}
-					errorIf(wErr, "Unable to read from HTTP body.")
-					writer.CloseWithError(wErr)
-					return
-				}
-				shaPayload := shaWriter.Sum(nil)
-				var s3Error APIErrorCode
-				if isRequestSignatureV4(r) {
-					s3Error = doesSignatureMatch(hex.EncodeToString(shaPayload), r, validateRegion)
-				} else if isRequestPresignedSignatureV4(r) {
-					s3Error = doesPresignedSignatureMatch(hex.EncodeToString(shaPayload), r, validateRegion)
-				}
-				var sErr error
-				if s3Error != ErrNone {
-					if s3Error == ErrSignatureDoesNotMatch {
-						sErr = errSignatureMismatch
-					} else {
-						sErr = fmt.Errorf("%v", getAPIError(s3Error))
-					}
-					writer.CloseWithError(sErr)
-					return
-				}
-				writer.Close()
-			}()
-
-			// Create object.
-			md5Sum, err = api.ObjectAPI.PutObject(bucket, object, size, reader, metadata)
-			// Close the pipe.
-			reader.Close()
-			// Wait for all the routines to finish.
-			wg.Wait()
+			reader = r.Body
+			signVerifyFn = signVerifyFactory(nil, r)
```

Comment:
https would mean we most of the times hit this code path, so even if the signature is wrong we end up reading the entire body. better would be:
```
if err:= signVerifyFactory(nil, r)(); err != nil {
    return
}
```